### PR TITLE
absolute new tiger

### PR DESCRIPTION
**Pull request by iskbaga**

### CODE REVIEW

File: modeling/models/NewTiger.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        return encoder_embs, encoder_mask, decoder_embs
+
+    def forward(self, inputs):
+        all_sample_events = inputs[
+            "{}.ids".format(self._sequence_prefix)
+        ]  # (all_batch_events)
+        all_sample_lengths = inputs[
+            "{}.length".format(self._sequence_prefix)
+        ]  # (batch_size)
+
```
Comment (NonameUntitled) [inline modeling/models/NewTiger.py]: all_sample_lengths
tgt_end_idx = torch.cumsum(all_sample_lengths, dim=-1)
tgt_start_idx = offsets - self._sem_id_length
flatten_mask = torch.zeros_like(all_sample_lengths)
flatten_mask[tgt_start_idx] += 1
flatten_mask[tgt_end_idx] -= 1
flatten_mask = torch.cumsum(flatten_mask, dim=-1).bool()
target_sem_ids = all_sample_events[flatten_mask]  # (batch_size * self._sem_id_length)

File: modeling/models/NewTiger.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        mask = range_row < enc_lens.unsqueeze(1)  # (batch_size, max_enc_len)
+
+        encoder_body = torch.zeros(batch_size, max_enc_len, emb_dimm, device=DEVICE)
+        encoder_body[mask] = sem_embs[:, :max_enc_len][mask]
+
+        pos_emb = self._get_position_embeddings(mask)
+        sem_pos_emb = self._get_sem_ids_position_embeddings(mask)
+        encoder_body += pos_emb + sem_pos_emb  # (batch_size, max_enc_len, embedding_dim)
+
+        bos = self.bos_embedding.view(1, 1, -1).expand(batch_size, 1, emb_dimm)
```
Comment (NonameUntitled) [inline modeling/models/NewTiger.py]: поменять bos на user

File: modeling/models/NewTiger.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+         decoder_input_embs) = (  # (batch_size, sem_id_len + 1, embedding_dim)
+            self.prepare_sem_id_batch(embeddings, all_sample_lengths)
+        )
+
+        if self.training:
+            after_encoder_emb, after_encoder_mask = self._apply_encoder(encoder_input_emb, encoder_input_mask)
+
+            # последние sem ids
+            # TODO ПРОВЕРИТЬ РАБОТОСПОСОБНОСТЬ, ЭТО ДОЛЖНО ВОЗВРАЩАТЬ ЧЕТЫРЕ ПОСЛЕДНИХ ВАЛИДНЫХ ТОКЕНА ВНУТРИ КАЖДОГО БАТЧА
+            events_2d = torch.zeros_like(mask, dtype=torch.long, device=DEVICE)
```
Comment (NonameUntitled) [inline modeling/models/NewTiger.py]: можно сделать маску, тут код неправильный

File: modeling/models/NewTiger.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                memory_key_padding_mask=~after_encoder_mask
+                # должно быть True для паддинга и False для реальных позиций
+            )  # (batch_size, sem_id_len, embedding_dim)
+
+            losses = []  # [(batch_size, codebook_size) * sem_id_len]
+            scores = []  # [(batch_size, codebook_size) * sem_id_len]
+            argmaxes = []  # [(batch_size, ) * sem_id_len]
+
+            for i in range(self._sem_id_len):
+                logits = torch.matmul(
```
Comment (NonameUntitled) [inline modeling/models/NewTiger.py]: добавить флажок "use_projection" +  nn.Layer если не обходимо
